app/models/user_notification.rb

```diff
@@ -12,13 +12,20 @@ class UserNotification < ApplicationRecord
 
   scope :ordered_by_updated_at, -> { order(updated_at: :asc) }
 
-  # The oldest unread popup notification for the given course and user.
+  # Returns the oldest unread popup notification for the given course user.
+  # Popups with deleted objects will be marked as read and ignored.
+  # +nil+ is returned if all popups are read.
   #
+  # @param [CourseUser] The course_user to check notifications for.
   # @return [UserNotification] The next popup notification to be shown.
   # @return [nil] if there are no unread notifications.
-  def self.next_unread_popup_for_course_user(course, user)
-    popup.where(user: user).unread_by(user).ordered_by_updated_at.
-      includes(activity: { object: :course }).
-      drop_while { |popup| popup.activity.from_course?(course) }.first
+  def self.next_unread_popup_for(course_user)
+    popup.where(user: course_user.user).ordered_by_updated_at.
+      includes(activity: { object: :course }).unread_by(course_user.user).
+      find do |popup|
+        present = popup.activity.object.present?
+        popup.mark_as_read! for: course_user.user unless present
```

Comment:
Why not delete the whole notification instead? No point keeping the invalid data around.
There'll be a record that an achievement/level was awarded which was then deleted but without the original achievement/level object that's not too useful anyway.